### ISSUE TITLE
The wrong results of AIA (average incremenatl accuracy) under half-base setting

### ISSUE THREAD

**Issue by Snowball0823** [simomagi/elastic_feature_consolidation#2]
As the accuracy per task you reported in the log, the AIA should be the mean of each task's Average Accuracy. However, the AIA in the summary.csv is not aligned with the mean of each task's Average Accuracy. The AIA I calculated from stdout.txt for CIFAR-100 under the hlaf-base setting is lower than 65%, but in the summary.csv, it shows the AIA is 69%. Could you verify and explain it? 

Here is the example:

In stdout.txt:
Computing New Task Prototypes

 >>> Test on task  0 : TAw acc= 85.0%, forg=  0.8% | TAg acc = 65.9%, forg = 19.9%  <<<

 >>> Test on task  1 : TAw acc= 91.6%, forg=  0.8% | TAg acc = 51.0%, forg = 20.4%  <<<

 >>> Test on task  2 : TAw acc= 94.6%, forg=  1.6% | TAg acc = 49.6%, forg = 23.6%  <<<

 >>> Test on task  3 : TAw acc= 97.0%, forg=  0.4% | TAg acc = 53.2%, forg = 22.2%  <<<

 >>> Test on task  4 : TAw acc= 97.2%, forg=  0.4% | TAg acc = 58.4%, forg = 19.4%  <<<

 >>> Test on task  5 : TAw acc= 95.6%, forg=  0.2% | TAg acc = 54.8%, forg = 26.2%  <<<

 >>> Test on task  6 : TAw acc= 89.8%, forg=  0.4% | TAg acc = 45.0%, forg = 28.8%  <<<

 >>> Test on task  7 : TAw acc= 85.2%, forg=  1.8% | TAg acc = 43.2%, forg = 27.2%  <<<

 >>> Test on task  8 : TAw acc= 88.8%, forg=  0.2% | TAg acc = 65.6%, forg = 11.4%  <<<

 >>> Test on task  9 : TAw acc= 93.0%, forg=  0.6% | TAg acc = 41.0%, forg = 36.8%  <<<

 >>> Test on task 10 : TAw acc= 92.4%, forg=  0.0% | TAg acc = 75.4%, forg =  0.0%  <<<

The Last_per_step_tag_acc -> (65.9+51.0+49.6+53.2+58.4+54.8+45.0+43.2+65.6+41.0+75.4)/11 = 54.87%

However, in summary.csv:

Last_per_step_tag_acc -> 59.83%

**Comment by simomagi** [comment]
Hi @Snowball0823,

Thank you for reaching out.

First of all, to avoid confusion and ensure consistency, I will use the same terminology adopted in our papers and in the logs (e.g., `summary.csv`) throughout this discussion. The accuracies reported in `stdout.txt` (65.9, 51.0, 49.6, ..., 75.4) refer to the accuracies on each task *after* learning the final task (task 10).

In the **Warm Start** setting (also known as the half-base setting), the first task is larger than the others. For example, in CIFAR-100 with 10 steps, the first task includes 50 classes, while each incremental step introduces 5 new classes. Therefore, in this setting, the **Last per Step Accuracy** (`Last_per_step_tag_acc` in `summary.csv`) is computed as:


 $$A_{step}^{10}=\frac{50 \cdot 65.9+ 5 \cdot 51.0+5 \cdot 49.6+5 \cdot 53.2+ 5 \cdot 58.4+5 \cdot 54.8+ 5 \cdot 45.0+ 5 \cdot 43.2+ 5 \cdot 65.6+ 5 \cdot 41.0+5 \cdot 75.4}{100}$$


This calculation matches the value reported in `summary.csv`. This evaluation protocol for the Warm Start setting is the one we follow and is also commonly used in other well-known continual learning frameworks (such as [FACIL](https://github.com/mmasana/FACIL)).

In the **Cold Start** setting (also known as equal task split), where each task has the same number of classes, your interpretation would indeed be correct, as each task contributes equally to the average.

Finally, it's worth clarifying that `summary.csv` also reports a different metric, called **Average Incremental Accuracy**, , which is distinct from the AIA metric you mentioned (that refers to our last per step accuracy). This metric is an alternative metric used to evaluate incremental learning approaches. The average incremental accuracy in the summary.csv is computed as:

$$A_{inc}^{10} =\frac{1}{11}* \sum_{i=0}^{10} A_{step}^i $$

In our extension paper [EFC++](https://arxiv.org/abs/2503.10439), (Section VI-C), we further discuss how these metrics are computed in both Warm and Cold Start scenarios.

I hope this clarifies the issue. I will update the README to provide further details on how the metrics are computed. Thank you again for reaching out.

**Comment by Snowball0823** [comment]
Hi @simomagi ,

Thank you for your explanation.

However, in the CIL, we always use AA (Average Accuracy) and AIA (Average Incremental Accuracy) as metrics, and also is commonly used in the popular framework (e.g. PyCIL). So, for a fair comparison, we use the final step's AA as a metric. In summary.csv (I believe that is the result that you reported in the paper), weighted average accuracy is used, which is not commonly used in other papers. 

According to the paper "A Comprehensive Survey of Continual Learning: Theory, Method and Application," even in the half-base setting, the AA is calculated as $AA_k=\frac{1}{k} \sum_{j=1}^k a_{k,j}$. The per-step incremental accuracy under the half-base setting would be impacted a lot by the first task's accuracy, so that it can not totally show the plasticity and stability, which are important for incremental learning. I would recommend reporting the AA and AIA for further comparison.

**Comment by simomagi** [comment]
Hi @Snowball0823, 
Thank you for the follow-up and for contributing to the discussion. I believe this clarification is very helpful for the community.

I have checked the survey  *A Comprehensive Survey of Continual Learning: Theory, Method and Application*,  as you suggested. I believe we are aligned in terms of metric definition. Specifically,  the **Average Incremental Accuracy** (AIA) is the same metric -- under the same name -- that we report in the summary.csv. The **Average Accuracy** (AA) defined in the survey corresponds to what we refer to as  **Last Per-Step Accuracy**.  Since, in the the **Cold Start** scenario all the tasks have the same number of classes, the formulas in Eq.(1) and (2) are exactly the ones we use to compute the two metrics.                                             

Now, about the **Warm Start** scenario (i.e. half-base split),   I may have  missed a discussion in the survey how these two metrics are computed in this specific setting. If you could point me to the relevant section in the paper, that would be very helpful. I agree with you that solely following the formulas in Eq. (1) and (2), the Warm Start setting should be treated as the Cold Start one.

That said, based on my understanding-- and comparison with implementations in repositories such as  PyCIL, PASS, FeTrIL, and FeCaM -- the metric is computed as us in the Warm Start setting, i.e. giving a more weight to the first task, albeit through a different implementation. 


I am not an expert of PyCIL, so please correct me if I'm wrong.  PyCIL appears to create a dynamic test_loader, that includes all samples up to the current task (this is common in other repositories too). As a result, when computing the total accuracy (AA),  the test samples from the first task have a larger influence on the final accuracy. I've manually verified this behavior, and what PyCIL refers to as Average Accuracy corresponds to our **Average Incremental Accuracy**, while the field  CNN:{"total":AA_k, ..} in their logs aligns with our **Last Per-Step Accuracy**,  assigning  increased weight on the large first task.  Our metric definitions are consistent with theirs, especially considering that test samples are typically balanced per class in common continual learning benchmarks (such CIFAR-100, Tiny-ImageNet, etc..). To show this equivalence mathematically, when computing the accuracy,  PyCIL effectively perform the following:   

![A_step](https://latex.codecogs.com/png.image?\dpi{130}&space;A_{step}^{K}&space;=&space;\frac{\sum_{i=0}^{K}&space;c^{K}_{i}}{\sum_{i=0}^{K}&space;N_i}&space;=&space;\frac{\sum_{i=0}^{K}&space;N_i&space;a_i^{K}}{\sum_{i=0}^{K}&space;N_i})

 
where $c^{K}_i$ is the number of correct samples of task $i$ after learning task $K$, $N_i$ is the total number of samples of task $i$ and $a_i^{K}$ is the accuracy on task $i$ after learning task $K$. Now, by denoting as $\vert \mathcal{C}_i \vert$ the number of classes in task $i$ and since each class as $n$ test samples (balanced test test), then $N_i = n \vert \mathcal{C}_i \vert$, this matches to our metric definition (section C in EFC++):

![A_step](https://latex.codecogs.com/png.image?\dpi{130}&space;A_{\text{step}}^{K}&space;=&space;\frac{\sum_{i=s}^{K}&space;\left\lvert\mathcal{C}_i\right\rvert&space;a_i^{K}}{\sum_{i=s}^{K}&space;\left\lvert\mathcal{C}_i\right\rvert})


We decided to report this metric for us and all the approaches,  to be much as fair as possible with what most incremental learning works perform when evaluating this scenario. However, I agree with your point  

*The per-step incremental accuracy under the half-base setting would be impacted a lot by the first task's accuracy, so that it can not totally show the plasticity and stability*. 

We acknowledge this in Section VI-C in [EFC++](https://arxiv.org/pdf/2503.10439): 

*It should be noted that in the Warm Start (WS) scenario, the performance on the initial classes, indicated as step
$s = 0$, is commonly included in the evaluation metric. Therefore, achieving higher stability on the large number of initial classes significantly impacts the final evaluation*

We stress this several times in paper, and I agree with you that the metric you propose can help to highlight better the plasticity. However, it does not reflect performance on the first task—**if** the first task is important, assigning it a higher weight is the appropriate choice. To explicitly show plasticity even in the Warm Start setting, we report the accuracy on each task in Fig. 6 and in the appendix in EFC++. Finally, if you would like to compute the metric you are suggesting, our code logs per-task accuracies in the acc_tag.out file (unweighted), following the convention used in FACIL. The average of the last row corresponds to the metric you're referring to.  

Thank you for the discussion—if you double-check what I'm saying in this discussion, it's much appreciated.

**Comment by Snowball0823** [comment]
Hi @simomagi ,

Thanks for your response and explanation!

Per your suggestion, I double-checked the code in the PyCIL and PASS repos. 

PyCIL: 
|
|___trainer.py ->

```python
for task in range(data_manager.nb_tasks):
        logging.info("All params: {}".format(count_parameters(model._network)))
        logging.info(
            "Trainable params: {}".format(count_parameters(model._network, True))
        )
        model.incremental_train(data_manager)
        cnn_accy, nme_accy = model.eval_task()
        ...
        cnn_curve["top1"].append(cnn_accy["top1"])
        cnn_curve["top5"].append(cnn_accy["top5"])

        logging.info("CNN top1 curve: {}".format(cnn_curve["top1"]))
        logging.info("CNN top5 curve: {}\n".format(cnn_curve["top5"]))

        print('Average Accuracy (CNN):', sum(cnn_curve["top1"])/len(cnn_curve["top1"]))
        logging.info("Average Accuracy (CNN): {}".format(sum(cnn_curve["top1"])/len(cnn_curve["top1"])))
```
Indeed, the `model.eval_task()` function directly calculates the whole accuracy. However, in the PASS,

PASS:
|
|___main.py ->

```python
print("############# Test for each Task #############")
for current_task in range(args.task_num+1):
    ...
    for i in range(current_task+1):
            if i == 0:
                # classes = [0, args.fg_nc]
                classes = whole_classes[0:args.fg_nc]
            else:
                # classes = [args.fg_nc + (i - 1) * task_size, args.fg_nc + i * task_size]
                classes = whole_classes[args.fg_nc + (i - 1) * task_size:args.fg_nc + i * task_size]            
            test_dataset.getTestData_up2now(classes, args.fg_nc + i * task_size)
            test_loader = DataLoader(dataset=test_dataset,
                                     shuffle=True,
                                     batch_size=args.batch_size)
            accuracy = correct.item() / total
            acc_up2now.append(accuracy)
    acc_all.append(acc_up2now)
    print(acc_up2now)
print(acc_all)
```
So it seems they report the $AA_k=\frac{1}{k} \sum_{j=1}^k a_{k,j}$ as AA.

It is confusing, but from my understanding, the incremental learning should focus on task performance. 

Furthermore, I want to clarify one thing. The Average Accuracy in the survey, or the first paper, iCaRL, only focuses on the tasks. In other words, no matter `cold start` or `warm start`, we should measure the performance of the method based on tasks. The only difference between the two settings is the number of classes in the first task. Thus, the metric of AA is defined based on the task, and it should not be weighted even in the `warm start` setting. Even in some scenarios, e.g, "if the first task is important," the incremental learning aims to fairly measure each task, so I think the $AA_k=\frac{1}{k} \sum_{j=1}^k a_{k,j}$ still can measure the first task as well.

I'm so glad that you pointed out that problem in EFC++. And thanks for the great discussion!

**Comment by simomagi** [comment]
Hi @Snowball0823,

Thanks again for the discussion.

Just for completeness, I went back and re-checked the PASS code. From what I see, they do compute the Warm Start accuracy in the same way as FACIL, PyCIL, and our codebase. It’s true that the snippet you pointed out stores the accuracy for each task separately, but if you look at the following lines 118-143 (main.py):

```python
print("############# Test for up2now Task #############")
test_dataset = iCIFAR100('./dataset', test_transform=test_transform, train=False, download=True)
test_dataset.targets = setup_data(test_dataset.targets, shuffle=True, seed=1993)
for current_task in range(args.task_num + 1):
    class_index = args.fg_nc + current_task * task_size
    ...
    classes = [0, args.fg_nc + current_task * task_size]
    test_dataset.getTestData_up2now(classes)  # accumulate samples up to the current task
    test_loader = DataLoader(dataset=test_dataset,
                             shuffle=True,
                             batch_size=args.batch_size)
    correct, total = 0.0, 0.0
    for step, (indexs, imgs, labels) in enumerate(test_loader): # in the last task, the test loader has all the tasks samples
        imgs, labels = imgs.to(device), labels.to(device)
        with torch.no_grad():
            outputs = model(imgs)
        outputs = outputs[:, ::4]
        predicts = torch.max(outputs, dim=1)[1]
        correct += (predicts.cpu() == labels.cpu()).sum()
        total += len(labels)
    accuracy = correct.item() / total
    print(accuracy)
```
We have a dynamic test loader, and in the last task the test loader contains all the samples from all the classes up to the current task. Indeed they compute the whole accuracy. This evaluation protocol appears to match that of PyCIL and FACIL and our codebase. I believe this is the metric they reported in the paper (I also remember running this code a few years ago).

That said, I fully agree that this kind of evaluation can obscure the model’s plasticity  in the Warm Start setting. So reporting the accuracy per task (via bar plot in EFC++) or the average (not weighted) as you suggested can allow to do this analysis. We tried to emphasize this point in our paper.

Thanks again for the great exchange! I’ll close this issue for now, but when I update the README to include details on the metric computation, I’ll link to this discussion.
